packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -236,52 +237,81 @@ async function resolveStaticMetadata(components: ComponentsType) {
 export async function collectMetadata(
   loaderTree: LoaderTree,
   props: any,
-  array: MetadataItems
+  array: MetadataItems,
+  isLeafLayout: boolean
 ) {
   const mod = await getLayoutOrPageModule(loaderTree)
   const staticFilesMetadata = await resolveStaticMetadata(loaderTree[2])
   const metadataExport = mod ? await getDefinedMetadata(mod, props) : null
 
-  array.push([metadataExport, staticFilesMetadata])
+  array.push({ metadataExport, staticFilesMetadata, isLeafLayout })
 }
 
 export async function accumulateMetadata(
   metadataItems: MetadataItems
 ): Promise<ResolvedMetadata> {
-  const deepClone: <T>(v: T) => T =
-    process.env.NODE_ENV === 'development'
-      ? require('next/dist/compiled/@edge-runtime/primitives/structured-clone')
-          .structuredClone
-      : (v) => v
   const resolvedMetadata = createDefaultMetadata()
 
-  let parentPromise = Promise.resolve(resolvedMetadata)
+  const resolvers: ((value: ResolvedMetadata) => void)[] = []
+  const generateMetadataResults: (Metadata | Promise<Metadata>)[] = []
 
-  for (const item of metadataItems) {
-    const [metadataExport, staticFilesMetadata] = item
-    const currentMetadata =
-      typeof metadataExport === 'function'
-        ? metadataExport(parentPromise)
-        : metadataExport
-    const layerMetadataPromise =
-      currentMetadata instanceof Promise
-        ? currentMetadata
-        : Promise.resolve(currentMetadata)
+  // call each `generateMetadata function concurrently and stash their resolver
+  for (let i = 0; i < metadataItems.length; i++) {
+    const { metadataExport } = metadataItems[i]
+    if (typeof metadataExport === 'function') {
+      generateMetadataResults.push(
+        metadataExport(
+          new Promise((resolve) => {
+            resolvers.push(resolve)
+          })
+        )
+      )
+    }
+  }
```

Comment:
It's a minor optimization but we could move this loop into the main one and start it's index from the first found 'function' export. the reason we may want to do this is for cases where there are no `generateMetadata` defined. In these cases we would only traverse the metadata items once instead of twice.